src/events/dns.js

```diff
@@ -0,0 +1,239 @@
+
+const initialDnsEvent = (name) => {
+    const startTime = Date.now();
+    const resource = new serverlessEvent.Resource([
+        'dns',
+        'dns',
+        name,
+    ]);
+    const dnsEvent = new serverlessEvent.Event([
+        `dns-${uuid4()}`,
+        utils.createTimestampFromTime(startTime),
+        null,
+        'dns',
+        0,
+        errorCode.ErrorCode.OK,
+    ]);
+    dnsEvent.setResource(resource);
+    return { dnsEvent, startTime };
+};
+
+const endDurationAndAddCallbackMetadata = (dnsEvent, startTime, err, metadata) => {
+    dnsEvent.setDuration(utils.createDurationTimestamp(startTime));
+    if (err) {
+        eventInterface.setException(dnsEvent, err);
+    } else {
+        eventInterface.addToMetadata(dnsEvent, { ...metadata });
+    }
+};
+
+const getRrtypeArguments = (arg1, arg2, arg3, functionName) => {
+    const hostname = arg1;
+    let rrtype = arg2;
+    let callback = arg3;
+    if (!arg3) {
+        if (functionName) {
+            rrtype = Object.values(rrtypesMethods).find(type => functionName.toLocaleLowerCase().includes((`query${type}`.toLocaleLowerCase())));
+        }
+        if (!rrtype) rrtype = rrtypesMethods.resolve4;
+        callback = arg2;
+    }
+    return { hostname, rrtype, callback };
+};
+
+const getArguments = (arg1, arg2, arg3) => {
+    const hostname = arg1;
+    let options = arg2;
+    let callback = arg3;
+    if (!arg3) {
+        options = undefined;
+        callback = arg2;
+    }
+    return { hostname, options, callback };
+};
+
+const buildParams = (arg1, arg2, arg3) => {
+    if (!arg3) {
+        return [arg1, arg2];
+    }
+    return [arg1, arg2, arg3];
+};
+
+const wrapDnsResolveFunction = original => (arg1, arg2, arg3) => {
+    let patchedCallback;
+    let clientRequest;
+    let options;
+    const { hostname, rrtype, callback } = getRrtypeArguments(arg1, arg2, arg3, original.name);
+    if (!callback) return original.apply(this, buildParams(arg1, arg2, arg3));
+    if (typeof arg2 === 'object') options = arg2;
+    try {
+        const { dnsEvent, startTime } = initialDnsEvent(original.name);
+        let requestArg2;
+        if (options) {
+            requestArg2 = { options };
+        } else {
+            requestArg2 = { 'Resource record type': rrtype };
+        }
+        eventInterface.addToMetadata(dnsEvent, { hostname, ...requestArg2 });
+        const responsePromise = new Promise((resolve) => {
+            patchedCallback = (err, arg) => {
+                let callbackArg;
+                if (options) {
+                    callbackArg = { addresses: arg };
+                } else {
+                    callbackArg = { records: arg };
+                }
+                endDurationAndAddCallbackMetadata(dnsEvent, startTime, err, { ...callbackArg });
+                resolve();
+                if (callback) callback(err, arg);
+            };
+        });
+        clientRequest = original.apply(this, [hostname, rrtype, patchedCallback]);
```

Comment:
just making sure that this always return a value even if callback was passed (otherwise you will invoke the original method twice). Please show it to me on sunday.